node/core/dispute-coordinator/src/tests.rs

```diff
@@ -3193,3 +3193,134 @@ fn participation_for_included_candidates() {
+			
+			// activate leaf - without candidate backed event
+			test_state
+				.activate_leaf_at_session(
+					&mut virtual_overseer,
+					session,
+					1,
+					vec![],
+				)
+				.await;
+
+			// generate two votes
+			let valid_vote = test_state
+				.issue_explicit_statement_with_index(
+					ValidatorIndex(1),
+					candidate_hash,
+					session,
+					true,
+				)
+				.await;
+
+			let invalid_vote = test_state
+				.issue_explicit_statement_with_index(
+					ValidatorIndex(2),
+					candidate_hash,
+					session,
+					false,
+				)
+				.await;
+
+			virtual_overseer
+				.send(FromOrchestra::Communication {
+					msg: DisputeCoordinatorMessage::ImportStatements {
+						candidate_receipt: candidate_receipt.clone(),
+						session,
+						statements: vec![
+							(valid_vote, ValidatorIndex(1)),
+							(invalid_vote, ValidatorIndex(2)),
+						],
+						pending_confirmation: None,
+					},
+				})
+				.await;
+
+			handle_approval_vote_request(&mut virtual_overseer, &candidate_hash, HashMap::new())
+				.await;
+
+			assert_matches!(virtual_overseer.recv().timeout(TEST_TIMEOUT).await, None);
+
+			// Step 2: Show that once backing votes are processed we participate
+
+			// Activate leaf: With candidate backed event
+			test_state
+				.activate_leaf_at_session(
+					&mut virtual_overseer,
+					session,
+					1,
+					vec![make_candidate_backed_event(candidate_receipt.clone())],
+				)
+				.await;
+
+			let backing_valid = test_state
+				.issue_backing_statement_with_index(ValidatorIndex(3), candidate_hash, session).await;
+
+			virtual_overseer
+				.send(FromOrchestra::Communication {
+					msg: DisputeCoordinatorMessage::ImportStatements {
+						candidate_receipt: candidate_receipt.clone(),
+						session,
+						statements: vec![
+							(backing_valid, ValidatorIndex(3))
+						],
+						pending_confirmation: None,
+					},
+				})
+				.await;
+			
+			participation_with_distribution(
+				&mut virtual_overseer,
+				&candidate_hash,
+				candidate_receipt.commitments_hash,
+			)
+			.await;
+
+			// Check for our 1 active dispute
+			let (tx, rx) = oneshot::channel();
+			virtual_overseer
+				.send(FromOrchestra::Communication {
+					msg: DisputeCoordinatorMessage::ActiveDisputes(tx),
+				})
+				.await;
+
+			assert_eq!(rx.await.unwrap().len(), 1);
+
+			// check if we have participated (casted a vote)
+			let (tx, rx) = oneshot::channel();
+			virtual_overseer
+				.send(FromOrchestra::Communication {
+					msg: DisputeCoordinatorMessage::QueryCandidateVotes(
+						vec![(session, candidate_hash)],
+						tx,
+					),
+				})
+				.await;
+
+			let (_, _, votes) = rx.await.unwrap().get(0).unwrap().clone();
+			assert_eq!(votes.valid.len(), 3); // 3 => 1 initial vote, 1 backing vote, and our vote
+			assert_eq!(votes.invalid.len(), 1);
+
+			// Wrap up
+			virtual_overseer.send(FromOrchestra::Signal(OverseerSignal::Conclude)).await;
+
+			test_state
+		})
+	});
+}
```

Comment:
What about a test that priority is bumped on inclusion events? This could be tested by queuing at least two participations, then bumping the one that would come later when both on the best-effort queue and then asserting that the "later" participation actually gets processed first.

Reply:
I had thought making such a test might be desirable, though I have a question about how to carry it out. Namely, it seems that `participation_with_distribution` would process both participation requests without recording which came first. Do you know how I can learn the ordering?

Reply:
`participation_with_distribution` does a single participation, if I'm not mistaken. So you can call it twice in the order you expect the things to happen.

Reply:
But won't such a test be a bit racy? When importing the events we'll start the participation immediately if we have got capacity.  So we can end up participating in the order enforced by the best effort queue before the included event manages to promote the dispute to priority?

Reply:
@eskimor @tdimitrov I'm trying to queue three participation requests by calling this three times:

`virtual_overseer
	.send(FromOrchestra::Communication {
		msg: DisputeCoordinatorMessage::ImportStatements {`

But the third `ImportStatements` message times out. This is solved if I `handle_approval_vote_request`s for the first two candidates right after importing statements. But if I try to `handle_approval_vote_request` for the third candidate right after importing, then things again blow up on me. 
Some support on this would be much appreciated, since it is shaping up to be one of the more complicated unit tests in all of disputes.

Reply:
It looks like we didn't build our testing functionality to be capable of handling queueing of multiple candidates at a time. You have to fully participate in each candidate before queueing another. Doing otherwise results in messages reaching the `virtual_overseer` in the wrong order, which breaks our test message handling. I could try to change our test helpers such as `handle_approval_vote_request` and `participation_with_distribution` to be less dependent on ordering, or perhaps there's a tricky workaround that could be implemented for this one test. I would love to learn that I'm wrong about this! Your thoughts would be much appreciated.

Reply:
Changing approach. Was trying to use the bare minimum number of candidates and to delay participating in any of them until all were in queued. As it turns out, disputes hates when you try to stall just one part of its functionality 😄